lightning/restore/restore.go

```diff
@@ -304,48 +305,254 @@ outside:
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
 
+func (worker *restoreSchemaWorker) doJob() {
+	var session checkpoints.Session
+	defer func() {
+		if session != nil {
+			session.Close()
+		}
+	}()
+loop:
+	for {
+		select {
+		case <-worker.ctx.Done():
+			// don't `return` or throw `worker.ctx.Err()`here,
+			// if we `return`, we can't mark cancelled jobs as done,
+			// if we `throw(worker.ctx.Err())`, it will be blocked to death
+			break loop
+		case job := <-worker.jobCh:
+			if job == nil {
+				// successful exit
+				return
+			}
+			var err error
+			if session == nil {
+				session, err = worker.glue.GetSession(worker.ctx)
+				if err != nil {
+					worker.wg.Done()
+					worker.throw(err)
+					// don't return
+					break loop
+				}
+			}
+			logger := log.With(zap.String("db", job.dbName), zap.String("table", job.tblName))
+			for _, stmt := range job.stmts {
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err = session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					err = errors.Annotatef(err, "%s %s failed", job.stmtType.String(), common.UniqueTable(job.dbName, job.tblName))
+					worker.wg.Done()
+					worker.throw(err)
+					// don't return
+					break loop
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+	// mark the cancelled job as `Done`, a little tricky,
+	// cauz we need make sure `worker.wg.Wait()` wouldn't blocked forever
+	for range worker.jobCh {
+		worker.wg.Done()
+	}
+}
 
+func (worker *restoreSchemaWorker) wait() error {
+	// avoid to `worker.wg.Wait()` blocked forever when all `doJob`'s goroutine exited.
+	// don't worry about goroutine below, it never become a zombie,
+	// cauz we have mechanism to clean cancelled jobs in `worker.jobCh`.
+	// means whole jobs has been send to `worker.jobCh` would be done.
+	waitCh := make(chan struct{})
+	go func() {
+		worker.wg.Wait()
+		close(waitCh)
+	}()
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return worker.ctx.Err()
+	case <-waitCh:
+		return nil
+	}
+}
+
+func (worker *restoreSchemaWorker) throw(err error) {
+	select {
+	case <-worker.ctx.Done():
+		// don't throw `worker.ctx.Err()` again, it will be blocked to death.
+		return
+	case worker.errCh <- err:
+		worker.quit()
+	}
+}
+
+func (worker *restoreSchemaWorker) appendJob(job *schemaJob) error {
+	worker.wg.Add(1)
+	select {
+	case err := <-worker.errCh:
+		// cancel the job
+		worker.wg.Done()
+		return err
+	case <-worker.ctx.Done():
+		// cancel the job
+		worker.wg.Done()
+		return worker.ctx.Err()
+	case worker.jobCh <- job:
+		return nil
+	}
+}
+
+func (rc *RestoreController) restoreSchema(ctx context.Context) error {
+	if !rc.cfg.Mydumper.NoSchema {
+		logTask := log.L().Begin(zap.InfoLevel, "restore all schema")
+		concurrency := 16
+		childCtx, cancel := context.WithCancel(ctx)
```

Comment:
should call `cancel` when we finish using this context. otherwise there's memory leak

Reply:
How about `cancel()` before line552 `err := worker.makeJobs(rc.dbMetas)`?

Reply:
Or `cancel()` in `makeJobs` defer function call?

Reply:
defer is enough

Reply:
Agree